src/main/java/org/folio/service/transactions/BatchTransactionService.java

```diff
@@ -81,4 +136,106 @@ private Future<Void> patchTransactions(List<TransactionPatch> patches, DBConn co
     return Future.failedFuture(new HttpException(500, "transactionPatches: not implemented"));
   }
 
+  private void populateMetadata(List<Transaction> transactions, Map<String, String> okapiHeaders) {
+    // NOTE: Okapi usually populates metadata when a POST method is used with an entity with metadata.
+    // But in the case of the batch API there is no top-level metadata, so it is not populated automatically.
+    for (Transaction tr : transactions) {
+      if (tr.getMetadata() == null) {
+        String userId = okapiHeaders.get("X-Okapi-User-Id");
+        if (userId == null) {
+          try {
+            userId = (new OkapiToken(okapiHeaders.get("X-Okapi-Token"))).getUserIdWithoutValidation();
+          } catch (Exception ignored) {
+            // could not find user id - ignoring
+          }
+        }
+        Metadata md = new Metadata();
+        md.setUpdatedDate(new Date());
+        md.setCreatedDate(md.getUpdatedDate());
+        md.setCreatedByUserId(userId);
+        md.setUpdatedByUserId(userId);
+        tr.setMetadata(md);
+      }
+    }
+  }
+
+  private Future<Void> createOrUpdate(List<Transaction> transactions,
+      BiFunction<TransactionService, Transaction, Future<Void>> createOrUpdateFct, DBConn conn) {
+    // Futures have to be executed in order, one after the other.
+    // Using functions instead of futures to avoid starting the futures before they are composed.
+    List<Function<Void, Future<Void>>> functions = new ArrayList<>();
+    groupByType(transactions).forEach((transactionType, transactionsForType) -> {
+      TransactionService transactionService = managingServiceFactory.findStrategy(transactionType);
+      groupBySummaryId(transactionsForType).forEach((summaryId, transactionsForSummary) -> {
+        if (TYPES_WITH_SUMMARIES.contains(transactionType)) {
+          functions.add(v -> createOrUpdateSummary(summaryId, transactionsForSummary, transactionType, conn));
+          functions.add(v -> removeTemporaryTransactions(summaryId, transactionType, conn));
+        }
+        transactionsForSummary.forEach(tr -> functions.add(v -> createOrUpdateFct.apply(transactionService, tr)));
+      });
+    });
+    Future<Void> future = Future.succeededFuture();
+    for (Function<Void, Future<Void>> fct : functions) {
+      future = future.compose(v -> fct.apply(null));
+    }
+    return future;
+  }
+
+  private Map<TransactionType, List<Transaction>> groupByType(List<Transaction> transactions) {
+    return transactions.stream().collect(groupingBy(tr -> tr.getTransactionType() == TransactionType.CREDIT ?
```

Comment:
if i'm not wrong here you mean that credit and payment will be in one group. In this case please add a comment about it. thank you

Reply:
Added a comment.

Reply:
I see the same logic in TransactionManagingStrategyFactory::findStrategy, can be useful to combine in some single place

Reply:
Yes